psga/common/primer_autodetection.nf

```diff
@@ -0,0 +1,81 @@
+/*
+ * Run: primer-autodetection (use first read only for illumina)
+ */
+process primer_autodetection {
+  publishDir "${params.output_path}/primer_autodetection", mode: 'copy', overwrite: true, pattern: '{*_primer.txt,*_primer_data.csv,*_primer_detection.tsv,trimmomatic.out,*.bowtie2*}'
+
+  tag "${task.index} - ${file_1}"
+
+  input:
+    path file_1
+    val kit
+
+  output:
+    path "*_primer.txt", emit: ch_primer
+    path "*_primer_data.csv", emit: ch_primer_data
+    path "*_primer_detection.tsv", emit: ch_primer_coverage
+    path "trimmomatic.out"
+    path "*.bowtie2*"
+
+  shell:
+  '''
+  kit=!{kit}
+  file_1=!{file_1}
+  sample_id=$( echo ${file_1} | cut -d '.' -f1 )
+  trimmed_sample="trimmed_sample.fastq"
+  crop="!{params.primer_minimum_length}"
+
+  # trim sample reads so that they are long as the shortest primer
+  java -jar /opt/trimmomatic/trimmomatic.jar SE -phred33 ${file_1} ${trimmed_sample} CROP:${crop} 2> trimmomatic.out
+
+  # iterate over the available primer scheme fasta.
+  # These fasta are generated by us and contain the sequences for each primer scheme name/version
+  for primer_fasta in $(find /primer_schemes -iname SARS-CoV-2.scheme.fasta); do
+      # extract the primer (e.g. ARTIC_V3)
+      primer="$(echo "${primer_fasta}" | cut -d/ -f3,5 | sed 's#/#_#')"
+
+      # build the index for the primer fasta
+      bowtie2-build ${primer_fasta} ${primer}.ref_idx &> ${primer}.bowtie2-build
+
+      # align the trimmed sample to the primer scheme fasta
+      bowtie2 --no-unal -x ${primer}.ref_idx -U ${trimmed_sample} 2> ${primer}.bowtie2 | samtools sort -o ${primer}.bam
+
+      # extract the coverage and other metrics from the bam. `samtools coverage` complains if the bam is not sorted
+      samtools coverage ${primer}.bam > ${primer}.coverage.tsv
+  done
+
+  # concatenate the coverage files to 1 single file with 1 single header
+  awk 'FNR==1 && NR!=1{next;}{print}' *.coverage.tsv > ${sample_id}_primer_detection.tsv
+
+  # extract 1) the primer with the highest number of reads and 2) the number of reads
+  detected_primer="$(awk 'BEGIN{nreads=0;primer="unknown"}{if(NR>1 && $4>nreads) {nreads=$4; primer=$1}} END {print primer,nreads}' ${sample_id}_primer_detection.tsv)"
```

Comment:
Expressions like this make it all very hard to understand. I think this whole script would be better of written in Python, it would also make it much easier to write tests for (and we definitely need unit tests for this)

Reply:
We can write a python script for this, but it will also mean that we will need to copy over the python script and install additional dependencies in the docker image.
What these awk commands do is pretty much the selection of a row or add two columns.